src/euler_types.jl

```diff
@@ -196,55 +196,78 @@ end
                              v[3])
 end
 
-@inline Base.:*(r1::RotZ, r2::RotZ) = RotZ(r1.theta + r2.theta)
 
-# define null rotations for convenience
-@inline eye(::Type{RotZ}) = RotZ(0.0)
-@inline eye{T}(::Type{RotZ{T}}) = RotZ{T}(zero(T))
+################################################################################
+################################################################################
 
-@inline inv(r::RotZ) = RotZ(-r.theta)
+######################
+# Two axis rotations #
+######################
 
-function Base.rand{R <: Union{RotX,RotY,RotZ}}(::Type{R})
+for axis1 in [:X, :Y, :Z]
+    Rot1Type = Symbol("Rot" * string(axis1))
+    for axis2 in filter(axis -> axis != axis1, [:X, :Y, :Z])
+        Rot2Type = Symbol("Rot" * string(axis2))
+        RotType = Symbol("Rot" * string(axis1) * string(axis2))
+        InvRotType = Symbol("Rot" * string(axis2) * string(axis1))
+
+        @eval begin
+            struct $RotType{T} <: Rotation{3,T}
+                theta1::T
+                theta2::T
+                $RotType{T}(theta1, theta2) where {T} = new{T}(theta1, theta2)
+                $RotType{T}(r::$RotType) where {T} = new{T}(r.theta1, r.theta2)
+            end
+
+            @inline $RotType(theta1::T1, theta2::T2) where {T1, T2} = $RotType{promote_type(T1, T2)}(theta1, theta2)
+            @inline $RotType(r::$RotType{T}) where {T} = $RotType{T}(r)
+
+            @inline convert(::Type{R}, r::$RotType) where {R<:$RotType} = R(r)
+            @inline convert(::Type{R}, r::R) where {R<:$RotType} = r
+
+            @inline function Base.getindex{T}(r::$RotType{T}, i::Int)
+                convert(Tuple, r)[i] # Slow...
+            end
+
+            @inline (::Type{R}){R<:$RotType}(t::NTuple{9}) = error("Cannot construct a two-axis rotation from a matrix")
+
+            # Composing single-axis rotations to obtain a two-axis rotation:
+            @inline Base.:*(r1::$Rot1Type, r2::$Rot2Type) = $RotType(r1.theta, r2.theta)
+
+            # Composing single-axis rotations with two-axis rotations:
+            @inline Base.:*(r1::$RotType, r2::$Rot2Type) = $RotType(r1.theta1, r1.theta2 + r2.theta)
+            @inline Base.:*(r1::$Rot1Type, r2::$RotType) = $RotType(r1.theta + r2.theta1, r2.theta2)
+
+            @inline inv(r::$RotType) = $InvRotType(-r.theta2, -r.theta1)
+
+            # define null rotations for convenience
+            @inline eye(::Type{$RotType}) = $RotType(0.0, 0.0)
+            @inline eye{T}(::Type{$RotType{T}}) = $RotType{T}(zero(T), zero(T))
+        end
+    end
+end
+
+function Base.rand{R <: Union{RotXY,RotYZ,RotZX, RotXZ, RotYX, RotZY}}(::Type{R})
     T = eltype(R)
     if T == Any
         T = Float64
     end
 
-    return R(2*pi*rand(T))
+    # Not really sure what this distribution is, but it's also not clear what
+    # it should be! rand(RotXY) *is* invariant to pre-rotations by a RotX and
+    # post-rotations by a RotY...
```

Comment:
Right, I guess this PR is probably not the best place to be removing things anyway.